### ISSUE TITLE
Feature request: ability to process detailed `ipcs -q` output

### ISSUE THREAD

**Issue by mailinglists35** [netdata/netdata#1890]
I have this output:
```
[root@linux bin]# ipcs -q

------ Message Queues --------
key        msqid      owner      perms      used-bytes   messages
0x00001bbc 0          root       666        0            0
0x00001c20 32769      root       666        0            0
0x00001c84 65538      root       666        0            0
0x00001ce8 98307      root       666        0            0
0x00001d4c 131076     root       666        0            0
0x00001db0 163845     root       666        0            0
0x000003e8 196614     root       666        0            0
0x000007d0 229383     root       666        0            0
0x00000bb8 262152     root       666        0            0
0x00000fa0 294921     root       666        0            0
0x00001388 327690     root       666        0            0
```

How do I see a graph for "messages" column for each message queue id and/or key? (used bytes could be optional. important pair is msqid + messages)

I know that there is the custom plugin but knowing my limited scripting abilities I'm pretty sure I'll create a messy bash script. Would love it this was natively supported.

Thank you!

**Comment by mailinglists35** [comment]
hello again :)

is there a chance you could revisit this sometime later on this year?

**Comment by ktsaou** [comment]
Yeap, I would like to have this feature too.
Let's plan it for v1.9.0

**Comment by l2isbad** [comment]
I see `ipcs` command provides info about:

- Shared memory segments
- Message queues
- Semaphore arrays

`Semaphore arrays` already [implemented in C](https://github.com/netdata/netdata/issues/1128#issuecomment-259550276) (i haven't found link to the actual commit/PR)

@ktsaou @vlvkobal 

I mean if we can avoid

**Comment by ktsaou** [comment]
> i think all others should be in C too.

yes, it is very simple to do it in C.
We just need to extend the ipc module of proc.plugin

https://github.com/netdata/netdata/blob/master/collectors/proc.plugin/ipc.c

**Comment by ktsaou** [comment]
I went through the code to check if calling `rrddim_free()` is safe.

Due to netdata's lockless design, the only thread expected to permanently have a pointer to a dimension is the data collection thread of the dimension. So, it should be safe to call `rrddim_free()` from that thread.

Keep in mind though that a lot of threads may be walking through the linked list of dimensions at that time. So, to delete the dimension, it is required to first write lock the chart by calling `rrdset_wrlock()`. This will ensure that no other thread is currently walking through the linked list of dimensions of the specific chart.

However, deleting a dimension like that is bad practice from a logical perspective. When you delete a dimension, its data are immediately lost. So, the user will not be able to review them. For short lived dimensions, it means the user will never have a chance to see them. The problem will also be worst if the same message queue is destroyed and re-created all the time - it will have no history at all. Finally, the streaming protocol does not support deleting dimensions, so the master will not delete them.

A better approach would be to add a flag to `RRDDIM` to mark the dimension as `OBSOLETE` (like we do for charts), and modify the plugins API to support this flag too. This will allow external plugins to set this flag and also streaming masters to receive it.

Then, the `OBSOLETE` dimensions should be deleted after some time (by default 1 hour - it is user configurable), via `rrdhost_cleanup_obsolete_charts()`.

**Comment by ktsaou** [comment]
This change is quite complex. It would require a lot of changes to various places (like the `OBSOLETE` flag for charts).

If you don't want to go through this, another solution would be:

- add a setting for collecting per message queue metrics and set it by default to `no`. So, people will enable it if they understand the implications.

or

- add a setting for the max number of message queue to be processed, let's say 200. if you add 200 dimensions to the chart, stop adding more.

For sure, the right solution is to support the `OBSOLETE` flag for dimensions too.

**Comment by ilyam8** [comment]
@vlvkobal this is has to be done sooner or later, i suggest to `spike` it

**Comment by Ferroin** [comment]
Tracking shared memory segments, OTOH, should in theory be pretty simple.  They can change wildly just like mqueues, but all that really matters is the count and the total size, so you're just counting entries and summing one of the columns in the output from `ipcs -m`.

**Comment by vlvkobal** [comment]
Implemented in #5115
